rust/src/transformations/manipulation/mod.rs

```diff
@@ -7,55 +7,96 @@ use opendp_derive::bootstrap;
 use crate::core::{Domain, Function, Metric, MetricSpace, StabilityMap, Transformation};
 use crate::domains::{AtomDomain, VectorDomain};
 use crate::error::*;
-use crate::metrics::{IntDistance, SymmetricDistance};
+use crate::metrics::{
+    ChangeOneDistance, HammingDistance, InsertDeleteDistance, IntDistance, SymmetricDistance,
+};
 use crate::traits::{CheckAtom, CheckNull, DistanceConstant};
 
+pub trait DatasetDomain: Domain {
+    type RowDomain: Domain;
+}
+
+impl<D: Domain> DatasetDomain for VectorDomain<D> {
+    type RowDomain = D;
+}
+
+pub trait DatasetMetric: Metric<Distance = IntDistance> {}
+
+impl DatasetMetric for SymmetricDistance {}
+impl DatasetMetric for InsertDeleteDistance {}
+impl DatasetMetric for ChangeOneDistance {}
+impl DatasetMetric for HammingDistance {}
+
+pub trait RowByRowDomain<DO: DatasetDomain>: DatasetDomain {
+    fn translate(&self, output_row_domain: DO::RowDomain) -> DO;
+    fn apply_rows(
+        value: &Self::Carrier,
+        row_function: &impl Fn(
+            &<Self::RowDomain as Domain>::Carrier,
+        ) -> Fallible<<DO::RowDomain as Domain>::Carrier>,
+    ) -> Fallible<DO::Carrier>;
+}
```

Comment:
Could you just fold this into `DatasetDomain`? That already has the concept of `RowDomain`, so doesn't seem like it'd limit generality.

Or maybe better would be to keep the separate traits, and move the `RowDomain` associated type down into `RowByRowDomain`?

Reply:
I see now that you need an associated type on `DatasetDomain` so that it's available for the output side, too. Still, it feels a little weird to call the associated type `RowDomain`. Something that takes the name `DatasetDomain` feels like it should be more general. What about datasets that don't consist of rows? Should it be something like `MemberDomain` or `ElementDomain`?

Reply:
Yes, it's a little strange with only one impl. I pulled this implementation out of the arrays PR because I liked the design. In that PR, `RowByRowDomain` is implemented for the combinations of `VectorDomain` and `Array2Domain`. We may have other viable combinations of input/output domains in the future.

Reply:
Renamed to `ElementDomain`.